flows/vision-zero/cris_import/cris_import.py

```diff
@@ -294,81 +320,246 @@ def remove_archives_from_sftp_endpoint(zip_location):
+
+        # Load up the list of imported records to iterate over. 
+        imported_records = util.load_input_data_for_keying(pg, DB_IMPORT_SCHEMA, table)
+
+        # Get columns used to uniquely identify a record
+        key_columns = mappings.get_key_columns()[output_map[table]]
+
+        # Build SQL fragment used as a JOIN ON clause to link input and output tables
+        linkage_clauses, linkage_sql = util.get_linkage_constructions(key_columns, output_map, table, DB_IMPORT_SCHEMA)
+
+        # Build list of columns available for import by inspecting the input table
+        input_column_names = util.get_input_column_names(pg, DB_IMPORT_SCHEMA, table, target_columns)
+
+        # iterate over each imported record and determine correct action
+        for source in imported_records:
+
+            # generate some record specific SQL fragments to identify the record in larger queries
+            public_key_sql, import_key_sql = util.get_key_clauses(table_keys, output_map, table, source, DB_IMPORT_SCHEMA)
+
+            # To decide to UPDATE, we need to find a matching target record in the output table.
+            # This function returns that record as a token of existence or false if none is available
+            if util.fetch_target_record(pg, output_map, table, public_key_sql):
+                # Build 3 arrays of SQL fragments, one element per column which can be `join`ed together in subsequent queries.
+                column_assignments, column_comparisons, column_aggregators = util.get_column_operators(target_columns, no_override_columns, source, table, output_map, DB_IMPORT_SCHEMA)
+
+                # Check if the proposed update would result in a non-op, such as if there are no changes between the import and
+                # target record. If this is the case, continue to the next record. There's no changes needed in this case.
+                if util.check_if_update_is_a_non_op(pg, column_comparisons, output_map, table, linkage_clauses, public_key_sql, DB_IMPORT_SCHEMA):
+                    print(f"Skipping update for {output_map[table]} {public_key_sql}")
+                    continue
+
+                # For future reporting and debugging purposes: Use SQL to query a list of 
+                # column names which have differing values between the import and target records.
+                # Return these column names as an array and display them in the output.
+                changed_columns = util.get_changed_columns(pg, column_aggregators, output_map, table, linkage_clauses, public_key_sql, DB_IMPORT_SCHEMA)
+                if len(changed_columns):
+                    print("Changed Columns: " + str(changed_columns["changed_columns"]))
+
+                # Using all the information we've gathered, form a single SQL update statement to update the target record.
+                update_statement = util.form_update_statement(output_map, table, column_assignments, DB_IMPORT_SCHEMA, public_key_sql, linkage_sql)
```

Comment:
TODO: This function should be extended to take `changed_columns` and query out the starting values of the columns in question in the input and target record. Printing these would be invaluable in tracking updates in the prefect logs.